README
NAME
    Math::Prime::XS - Detect and calculate prime numbers with deterministic
    tests

SYNOPSIS
     use Math::Prime::XS ':all';
     # or
     use Math::Prime::XS qw(is_prime primes mod_primes sieve_primes sum_primes trial_primes);

     print "prime" if is_prime(59);

     @all_primes   = primes(100);
     @range_primes = primes(30, 70);

     @all_primes   = mod_primes(100);
     @range_primes = mod_primes(30, 70);

     @all_primes   = sieve_primes(100);
     @range_primes = sieve_primes(30, 70);

     @all_primes   = sum_primes(100);
     @range_primes = sum_primes(30, 70);

     @all_primes   = trial_primes(100);
     @range_primes = trial_primes(30, 70);

DESCRIPTION
    "Math::Prime::XS" detects and calculates prime numbers by either
    applying Modulo operator division, the Sieve of Eratosthenes, a
    Summation calculation or Trial division.

FUNCTIONS
  is_prime
     is_prime($number);

    Returns true if the number is prime, false if not.

    The XS function invoked within "is_prime()" is subject to change
    (currently it's an all-XS trial division skipping multiples of 2,3,5).

  primes
     @all_primes   = primes($number);
     @range_primes = primes($base, $number);

    Returns all primes for the given number or primes between the base and
    number.

    The resolved function called is subject to change (currently
    "sieve_primes()").

  mod_primes
     @all_primes   = mod_primes($number);
     @range_primes = mod_primes($base, $number);

    Applies the Modulo operator division algorithm:

    Divide the number by 2 and all odd numbers <= sqrt(n); if any divides
    exactly then the number is not prime.

    Returns all primes between 2 and $number, or between $base and $number
    (inclusive).

    (This function differs from "trial_primes" in that the latter takes some
    trouble to divide only by primes below sqrt(n), whereas "mod_primes"
    divides by all integers not easily identifiable as composite.)

  sieve_primes
     @all_primes   = sieve_primes($number);
     @range_primes = sieve_primes($base, $number);

    Applies the Sieve of Eratosthenes algorithm:

    One of the most efficient ways to find all the small primes (say, all
    those less than 10,000,000) is by using the Sieve of Eratosthenes (ca
    240 BC). Make a list of all numbers less than or equal to n (and greater
    than one) and strike out the multiples of all primes less than or equal
    to the square root of n: the numbers that are left are primes.

    Returns all primes for the given number or primes between the base and
    number.

    <http://primes.utm.edu/glossary/page.php?sort=SieveOfEratosthenes>

  sum_primes
     @all_primes   = sum_primes($number);
     @range_primes = sum_primes($base, $number);

    Applies the Summation calculation algorithm:

    The summation calculation algorithm resembles the modulo operator
    division algorithm, but also shares some common properties with the
    Sieve of Eratosthenes. For each saved prime smaller than or equal to the
    square root of the number, recall the corresponding sum (if none, start
    with zero); add the prime to the sum being calculated while the
    summation is smaller than the number. If none of the sums equals the
    number, then the number is prime.

    Returns all primes for the given number or primes between the base and
    number.

    <http://www.geraldbuehler.de/primzahlen/>

  trial_primes
     @all_primes   = trial_primes($number);
     @range_primes = trial_primes($base, $number);

    Applies the Trial division algorithm:

    To see if an individual small number is prime, trial division works
    well: just divide by all the primes less than or equal to its square
    root. For example, to assert 211 is prime, divide by 2, 3, 5, 7, 11 and
    13. Since none of these primes divides the number evenly, it is prime.

    Returns all primes for the given number or primes between the base and
    number.

    <http://primes.utm.edu/glossary/page.php?sort=TrialDivision>

BENCHMARK
    Following output resulted from a benchmark measuring the time to
    calculate primes up to 1,000,000 with 100 iterations for each function.
    The tests were conducted by the "cmpthese" function of the Benchmark
    module.

                    Rate   mod_primes trial_primes   sum_primes sieve_primes
     mod_primes   1.32/s           --         -58%         -79%         -97%
     trial_primes 3.13/s         137%           --         -49%         -93%
     sum_primes   6.17/s         366%          97%           --         -86%
     sieve_primes 43.3/s        3173%        1284%         602%           --

    The "Rate" column is the speed in how many times per second, so
    "sieve_primes()" is the fastest for this particular test.

EXPORT
  Functions
    "is_prime(), primes(), mod_primes(), sieve_primes(), sum_primes(),
    trial_primes()" are exportable.

  Tags
    ":all - *()"

BUGS & CAVEATS
    Note that the order of execution speed for functions may differ from the
    benchmarked results when numbers get larger or smaller.

SEE ALSO
    <http://primes.utm.edu>,
    <http://www.it.fht-esslingen.de/~schmidt/vorlesungen/kryptologie/seminar
    /ws9798/html/prim/prim-1.html>

AUTHOR
    Steven Schubiger <schubiger@cpan.org>

LICENSE
    This program is free software; you may redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://dev.perl.org/licenses/>



Changes
Revision history for Perl extension Math::Prime::XS.

0.26_01  2013-04-28  <kryde@cpan.org>

 - New count_primes() as suggested by Dana Jacobsen.

0.26  2011-11-24  <kryde@cpan.org>

 - Stable release.

0.25_02  2011-10-30  <kryde@cpan.org>

 - Fix functions.t test count for 64-bit long

0.25_01  2011-10-25  <kryde@cpan.org>

 - is_prime() all-XS code for speedup
 - Try Perl's TRUE/FALSE to avoid stdbool.h configury

0.25  2011-09-24  <kryde@cpan.org>

 - Oops, restore stdbool.h bit from 0.23_01.

0.24  2011-09-23  <kryde@cpan.org>

 - mod_primes() fix: avoid infinite loop if number==2**32-1
 - mod_primes() enhancement: consider only odd integers from "base"
   upwards; divide only by odd numbers up to sqrt(n)

0.23_01  2011-09-22  <schubiger@cpan.org>

 - Fix build failure caused by perl's inclusion of stdbool.h.
   [perl #99408 - Andreas Koenig]

0.23  2011-01-10  <schubiger@cpan.org>

 - Merged development version to stable.

0.22_01  2010-12-07  <schubiger@cpan.org>

 - Fix overflow in xs_sieve_primes() marking 1928099 as composite
   when it's not and add a test. [rt #62632 - Kevin Ryde]

 - Strike out multiples in xs_sieve_primes() up to the square root
   of the number.

 - Define and use a macro to test for evenness.

0.22  2010-10-24  <schubiger@cpan.org>

 - Merged development version to stable.

0.21_01  2010-10-15  <schubiger@cpan.org>

 - Omit checking for validations if the base equals the number.
   [rt #62112 - Kevin Ryde]

 - Add tests for calling the prime calculating functions with
   the same base and number.

 - Remove declaration of an unused variable in xs_sieve_primes().

0.21  2010-10-08  <schubiger@cpan.org>

 - Merged development version to stable.

0.20_02  2010-10-07  <schubiger@cpan.org>

 - Allocate memory dynamically for xs_*_primes() instead of using
   fixed-width arrays. [rt #58322 - Kevin Ryde]

 - Utilize the composite array in xs_sieve_primes() as bit field.

 - Don't use a hash in xs_trial_primes() to save primes.
   [rt #58074 - Kevin Ryde]

 - Optimize the implementation of some algorithms.

 - Store numbers with automatically growing their memory.

 - Separate the base number initialization and check from the XSUBs.

 - is_prime() wraps around a xs_*_primes XSUB.

 - Remove xs_is_prime() as it's no longer used.

 - Validation of arguments is handled by Params::Validate.

 - Use the boolean pragma for true/false values.

 - Document both invocations of each prime calculating function.

 - Refine the documentation of these functions and is_prime().

 - Renew the benchmark results and rephrase its description.

 - Mention bugs and caveats.

 - Adapt the synopsis to use greater numbers.

 - Reword the abstract.

 - Update broken license link.

 - Be more verbose while testing by printing the numbers.

 - Remove the testing diagnostic.

 - Rename calc_primes.t to functions.t.

 - Adjust test file permissions.

0.20_01 Thu May 15 17:13:07 CEST 2008

 - Use memset() within xs_sieve_primes() to "clear" the
   composite array.

0.20 Wed May 14 15:46:42 CEST 2008

 - Improved the speed of xs_sieve_primes() by using a fixed-width
   array for the composites, instead of a hash.
   [suggested by Moritz Lenz <http://search.cpan.org/~moritz>]

 - Reformatted the source code layout of XS.xs.

0.19 Sat Feb 23 16:20:29 CET 2008

 - Use XSRETURN_IV() instead of manually extending the stack
   and pushing the mortal boolean number.

0.18 Fri Feb 15 15:23:36 CET 2008

 - is_prime() now returns 0 if the number isn't a prime number.

 - Added tests for calculating primes within a range.

 - Updated ppport.h.

0.17 Tue Apr  4 20:53:19 CEST 2006

 - Changed Build.PL argument create_makefile_pl from 'passtrough'
   to 'traditional' for the sake of compatibility.

0.15 Thu Nov  3 20:45:22 CET 2005

 - Omitted the pod-coverage testing part for XSUBs.

0.14 Sat Oct 29 20:30:42 CEST 2005

 - Added INSTALL instructions to the package.

 - The xs_files argument in Build.PL specifies where *.xs
   files reside.

0.13 Wed Oct 26 19:44:23 CEST 2005

 - Converted Math::Prime::Simple over to XS.

 - Reimplemented the Sieve of Eratosthenes in C.

 - Added modulo operator division, trial division and a
   "hybrid" summing calculation method.

 - Updated the docs with relevant function documentation
   and related benchmarks.

0.03 2004/01/14

 - Implemented the Sieve of Eratosthenes.

0.01 2004/01/14

 - Initial version.


devel/cmpthese.pl
#!/usr/bin/perl -w

use 5.004;
use Benchmark 'cmpthese';
use Math::Prime::XS ':all';

print "Math::Prime::XS version ",Math::Prime::XS->VERSION,"\n";

$number = 1000003; # a prime
cmpthese(100000, {
             mod_primes => sub { mod_primes($number,$number) },
             # sum_primes => sub { sum_primes($number,$number) },
             # trial_primes => sub { trial_primes($number,$number) },
             # sieve_primes => sub { sieve_primes($number,$number) },
            });

my $number = 1_000_000;
cmpthese(100, {
             mod_primes => sub { mod_primes($number) },
             sum_primes => sub { sum_primes($number) },
             trial_primes => sub { trial_primes($number) },
             sieve_primes => sub { sieve_primes($number) },
            });

exit 0;









# version 0.23 was
#
#                    Rate   mod_primes   sum_primes trial_primes sieve_primes
# mod_primes   6.39e-03/s           --        -100%        -100%        -100%
# sum_primes       8.10/s      126540%           --         -41%         -89%
# trial_primes     13.6/s      212979%          68%           --         -82%
# sieve_primes     74.6/s     1167064%         822%         448%           --
#
# and on Kevin's pc roughly
#
#                    Rate   mod_primes trial_primes   sum_primes sieve_primes
# mod_primes   1.26e-03/s           --        -100%        -100%        -100%
# trial_primes     3.12/s      247819%           --         -53%         -94%
# sum_primes       6.67/s      528793%         113%           --         -87%
# sieve_primes     50.0/s     3966600%        1500%         650%           --


lib/Math/Prime/XS.pm
package Math::Prime::XS;

use strict;
use warnings;
use base qw(Exporter);
use boolean qw(true false);

use Carp qw(croak);
use Params::Validate ':all';
use Scalar::Util qw(looks_like_number);

our ($VERSION, @EXPORT_OK, %EXPORT_TAGS);
my @subs;

$VERSION = '0.26_01';
@subs = qw(is_prime primes count_primes
           mod_primes sieve_primes sum_primes trial_primes);
@EXPORT_OK = @subs;
%EXPORT_TAGS = ('all' => [ @subs ]);

validation_options(
    on_fail => sub
{
    my ($error) = @_;
    chomp $error;
    croak $error;
},
    stack_skip => 2,
);

*primes = \&sieve_primes;
sub count_primes { _validate(@_); @_ == 1 ? xs_sieve_count_primes($_[0], 2) : xs_sieve_count_primes(reverse @_) }

# Reverse arguments for xs_*_primes() when both base and number are specified
sub mod_primes   { _validate(@_); @_ == 1 ? xs_mod_primes  ($_[0], 2) : xs_mod_primes  (reverse @_) }
sub sieve_primes { _validate(@_); @_ == 1 ? xs_sieve_primes($_[0], 2) : xs_sieve_primes(reverse @_) }
sub sum_primes   { _validate(@_); @_ == 1 ? xs_sum_primes  ($_[0], 2) : xs_sum_primes  (reverse @_) }
sub trial_primes { _validate(@_); @_ == 1 ? xs_trial_primes($_[0], 2) : xs_trial_primes(reverse @_) }

sub _validate
{
    my $positive_num = sub { looks_like_number($_[0]) && $_[0] >= 0 };

    validate_pos(@_,
        { type => SCALAR,
          callbacks => {
            'is a positive number' => $positive_num,
          },
        },
        { type => SCALAR,
          optional => true,
          callbacks => {
            'is a positive number' => $positive_num,
          },
        },
    );
    if (@_ == 2) {
        my ($base, $number) = @_;
        croak 'Base is greater than the number' if $base > $number;
    }
}

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

1;
__END__

=head1 NAME

Math::Prime::XS - Detect and calculate prime numbers with deterministic tests

=head1 SYNOPSIS

 use Math::Prime::XS ':all';
 # or
 use Math::Prime::XS qw(is_prime primes mod_primes sieve_primes sum_primes trial_primes);

 print "prime" if is_prime(59);

 @all_primes   = primes(100);
 @range_primes = primes(30, 70);

 @all_primes   = mod_primes(100);
 @range_primes = mod_primes(30, 70);

 @all_primes   = sieve_primes(100);
 @range_primes = sieve_primes(30, 70);

 @all_primes   = sum_primes(100);
 @range_primes = sum_primes(30, 70);

 @all_primes   = trial_primes(100);
 @range_primes = trial_primes(30, 70);

=head1 DESCRIPTION

C<Math::Prime::XS> detects and calculates prime numbers by either applying
Modulo operator division, the Sieve of Eratosthenes, a Summation calculation
or Trial division.

=head1 FUNCTIONS

=head2 is_prime

 is_prime($number);

Returns true if the number is prime, false if not.

The XS function invoked within C<is_prime()> is subject to change (currently
it's an all-XS trial division skipping multiples of 2,3,5).

=head2 primes

 @all_primes   = primes($number);
 @range_primes = primes($base, $number);

Returns all primes for the given number or primes between the base and number.

The resolved function called is subject to change (currently C<sieve_primes()>).

=head2 count_primes

 $count = count_primes($number);
 $count = count_primes($base, $number);

Return a count of primes from 0 to C<$number>, or from C<$base> to
C<$number>, inclusive.  The arguments are the same as C<primes()> but the
return is just a count of the primes.

=head1 SPECIFIC ALGORITHMS

=head2 mod_primes

 @all_primes   = mod_primes($number);
 @range_primes = mod_primes($base, $number);

Applies the Modulo operator division algorithm:

Divide the number by 2 and all odd numbers E<lt>= sqrt(n); if any divides
exactly then the number is not prime.

Returns all primes between 2 and C<$number>, or between C<$base> and
C<$number> (inclusive).

(This function differs from C<trial_primes> in that the latter takes some
trouble to divide only by primes below sqrt(n), whereas C<mod_primes>
divides by all integers not easily identifiable as composite.)

=head2 sieve_primes

 @all_primes   = sieve_primes($number);
 @range_primes = sieve_primes($base, $number);

Applies the Sieve of Eratosthenes algorithm:

One of the most efficient ways to find all the small primes (say, all those less
than 10,000,000) is by using the Sieve of Eratosthenes (ca 240 BC). Make a list
of all numbers less than or equal to n (and greater than one) and strike out
the multiples of all primes less than or equal to the square root of n:
the numbers that are left are primes.

Returns all primes for the given number or primes between the base and number.

L<http://primes.utm.edu/glossary/page.php?sort=SieveOfEratosthenes>

=head2 sum_primes

 @all_primes   = sum_primes($number);
 @range_primes = sum_primes($base, $number);

Applies the Summation calculation algorithm:

The summation calculation algorithm resembles the modulo operator division
algorithm, but also shares some common properties with the Sieve of
Eratosthenes. For each saved prime smaller than or equal to the square root
of the number, recall the corresponding sum (if none, start with zero);
add the prime to the sum being calculated while the summation is smaller
than the number. If none of the sums equals the number, then the number
is prime.

Returns all primes for the given number or primes between the base and number.

L<http://www.geraldbuehler.de/primzahlen/>

=head2 trial_primes

 @all_primes   = trial_primes($number);
 @range_primes = trial_primes($base, $number);

Applies the Trial division algorithm:

To see if an individual small number is prime, trial division works well:
just divide by all the primes less than or equal to its square root. For
example, to assert 211 is prime, divide by 2, 3, 5, 7, 11 and 13. Since
none of these primes divides the number evenly, it is prime.

Returns all primes for the given number or primes between the base and number.

L<http://primes.utm.edu/glossary/page.php?sort=TrialDivision>

=head1 BENCHMARK

Following output resulted from a benchmark measuring the time to calculate
primes up to 1,000,000 with 100 iterations for each function. The tests
were conducted by the C<cmpthese> function of the Benchmark module.

                Rate   mod_primes trial_primes   sum_primes sieve_primes
 mod_primes   1.32/s           --         -58%         -79%         -97%
 trial_primes 3.13/s         137%           --         -49%         -93%
 sum_primes   6.17/s         366%          97%           --         -86%
 sieve_primes 43.3/s        3173%        1284%         602%           --

The "Rate" column is the speed in how many times per second, so
C<sieve_primes()> is the fastest for this particular test.

=head1 EXPORT

=head2 Functions

C<is_prime(), primes(), mod_primes(), sieve_primes(), sum_primes(), trial_primes()>
are exportable.

=head2 Tags

C<:all - *()>

=head1 BUGS & CAVEATS

Note that the order of execution speed for functions may differ from the
benchmarked results when numbers get larger or smaller.

=head1 SEE ALSO

L<http://primes.utm.edu>,
L<http://www.it.fht-esslingen.de/~schmidt/vorlesungen/kryptologie/seminar/ws9798/html/prim/prim-1.html>

=head1 AUTHOR

Steven Schubiger <schubiger@cpan.org>

=head1 LICENSE

This program is free software; you may redistribute it and/or
modify it under the same terms as Perl itself.

See L<http://dev.perl.org/licenses/>

=cut
